File: Data Structures & Algorithms/redundant-connection/submission-1.java
class Solution {

    public int[] findRedundantConnection(int[][] edges) {
        int[] nodes = new int[edges.length + 1];

        for (int i = 1; i < nodes.length; i++) {
            nodes[i] = i;
        }

        for (int[] edge : edges) {
            int n1Root = nodes[edge[0]];
            int n2Root = nodes[edge[1]];

            if (n1Root == n2Root) {
                return edge;
            }

            union(nodes, n1Root, n2Root);
        }

        return null;
    }

    private void union(int[] nodes, int n1, int n2) {
        for (int i = 0; i < nodes.length; i++) {
            if (nodes[i] == n1) {
                nodes[i] = n2;
            }
        }
    }
}

File: Data Structures & Algorithms/count-connected-components/submission-1.kt
class Solution {
    fun countComponents(n: Int, edges: Array<IntArray>): Int {
        var unionFind = UnionFind(n)

        for (edge in edges) {
            unionFind.union(edge[0], edge[1])
        }

        return unionFind.getUniqueCount()
    }

    private class UnionFind(n: Int) {

        private val arr = IntArray(n)

        init {
            for (i in 0 until n) {
                arr[i] = i
            }
        }

        fun getUniqueCount(): Int {
            return arr.distinct().size
        }

        fun union(v1: Int, v2: Int) {
            val v1Root = find(v1)
            val v2Root = find(v2)

            for (i in 0 until arr.size) {
                if (arr[i] == v2Root)
                    arr[i] = v1Root
            }
        }

        private fun find(v: Int): Int {
            if (arr[v] == v) return v

            return find(arr[v])
        }
    }
}